Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Date:  20200903 (Unix timestamp 1599195647)
Summary:  Replace buggy  patch -R  with sandboxed copy.

patch -R does not work well with combinediff patches with split patches
combined in one file of the same filename instead of one.  Using patch -R may
complain about Unreversed patch detected!  Ignore -R? [n].

v2:  restore only affected files
v1:  copy entire SRCDIR

diff -urp kpatch-0.9.1.orig/kpatch-build/kpatch-build kpatch-0.9.1/kpatch-build/kpatch-build
--- kpatch-0.9.1.orig/kpatch-build/kpatch-build	2020-09-03 23:25:11.098271942 -0700
+++ kpatch-0.9.1/kpatch-build/kpatch-build	2020-09-03 23:31:37.402973699 -0700
@@ -43,6 +43,7 @@ ARCH="$(uname -m)"
 CPUS="$(getconf _NPROCESSORS_ONLN)"
 CACHEDIR="${CACHEDIR:-$HOME/.kpatch}"
 SRCDIR="$CACHEDIR/src"
+PATCH_SANDBOX="$CACHEDIR/patchsandbox"
 RPMTOPDIR="$CACHEDIR/buildroot"
 VERSIONFILE="$CACHEDIR/version"
 TEMPDIR="$CACHEDIR/tmp"
@@ -119,6 +120,12 @@ verify_patch_files() {
 apply_patches() {
 	local patch
 
+	mkdir -p "${PATCH_SANDBOX}"
+
+	for f in $(cat "${PATCH_LIST[@]}" | grep -E -e "^[-][-][-]" | cut -f 2- -d "/" | sort | uniq | sed -e "/dev[/]null/d") ; do
+		cp -a "$SRCDIR/${f}" "${PATCH_SANDBOX}"
+	done
+
 	for patch in "${PATCH_LIST[@]}"; do
 		if [[ "${DIE_ON_PATCH_DRY_RUN}" == "1" ]] ; then
 			patch -N -p1 --dry-run < "$patch" 2>&1 | logger || die \
@@ -138,11 +145,7 @@ remove_patches() {
 	local patch
 	local idx
 
-	for (( ; APPLIED_PATCHES>0; APPLIED_PATCHES-- )); do
-		idx=$(( APPLIED_PATCHES - 1))
-		patch="${PATCH_LIST[$idx]}"
-		patch -p1 -R -d "$SRCDIR" < "$patch" &> /dev/null
-	done
+	cp -a "${PATCH_SANDBOX}" "${SRCDIR}"
 
 	# If $SRCDIR was a git repo, make sure git actually sees that
 	# we've reverted our patch(es).
